src/lsm/zig_zag_merge.zig

```diff
@@ -207,8 +205,22 @@ pub fn ZigZagMergeIteratorType(
                 }
             }
 
+            // The iterator cannot regress.
+            assert(it.probe_key_previous == null or
+                probe_key == it.probe_key_previous.? or
+                it.key_ahead(probe_key, it.probe_key_previous.?));
+
+            it.probe_key_previous = probe_key;
             return if (drained.count() == 0) probe_key else error.Drained;
         }
+
+        /// Depending on the direction, returns true if key `a` is ahead of key `b`.
+        inline fn key_ahead(it: *const ZigZagMergeIterator, a: Key, b: Key) bool {
```

Comment:
Nit, but maybe change this to take a `struct { key_before, key_after }`, since I think the argument order is confusing otherwise.